[Dynamic Fields] Replace Child Object docs with Dynamic Field Docs (#5447)
MystenLabs/sui: doc/src/build/programming-with-objects/ch6-collections.md

### Current Limitations

Collections are built on top of [dynamic fields](ch5-dynamic-fields.md), and so are subject to its [limitations](ch5-dynamic-fields.md#current-limitations). Additionally, the following functionality is planned, but not currently supported:

- `sui::bag::contains<K: copy + drop + store>(bag: &Bag, k: K): bool` which checks whether a key-value pair exists in `bag` with key `k: K` and a value of any type (in addition to `contain_with_type` which performs a similar check, but requires passing a specific value type).

### Tables

```rust
module sui::table {

struct Table<K: copy + drop + store, V: store> has key, store { /* ... */ }

public fun new<K: copy + drop + store, V: store>(
    ctx: &mut TxContext,
): Table<K, V>;

}
```

`Table<K, V>` is a *homogeneous* map, meaning that all its keys have the same type as each other (`K`), and all its values have the same type as each other as well (`V`). It is created with `sui::table::new`, which requires access to a `&mut TxContext` because `Table`s are objects themselves, which can be transfered, shared, wrapped, or unwrapped, just like any other object.

> > :bulb: As usual, there is an equivalent `sui::object_table::ObjectTable` with an additional `V: key + store` constraint.

Reviewer: ```suggestion
> :bulb: As with `dynamic_field` and `dynamic_object_field`, there is an equivalent `sui::object_table::ObjectTable` with an additional `V: key + store` constraint.
```

### Bags

```rust
module sui::bag {

struct Bag has key, store { /* ... */ }

public fun new(ctx: &mut TxContext): Bag;

}
```

`Bag` is a *heterogeneous* map, so it can hold key-value pairs of arbitrary types (they don't need to match each other). Note that the `Bag` type does not have any type parameters for this reason. Like `Table`, `Bag` is also an object, so creating one with `sui::bag::new` requires supplying a `&mut TxContext` to generate an ID.

> > :bulb: See `sui::bag::ObjectBag` for the object-preserving version of `Bag`.

Reviewer: I like this wording a lot. Maybe replicate it for the `table` bulb above?

### Querying for Containment

All collections can be queried for key containment with:

```rust
module sui::table {

public fun contains<K: copy + drop + store, V: store>(
    table: &Table<K, V>
    k: K
): bool;

}
```

The equivalent function for `Bag` is,

```rust
module sui::bag {

public fun contains_with_type<K: copy + drop + store, V: store>(
    bag: &Bag,
    k: K
): bool;

}
```

which tests whether `bag` contains a key-value pair with key `k: K` and some value of type `V`.

### Clean-up

As mentioned in the introduction, collection types protect against accidental deletion when they might not be empty. This protection comes from the fact that they do not have `drop`, so must be explicitly deleted, using this API:

```rust
module sui::table {

public fun destroy_empty<K: copy + drop + store, V: store>(
    table: Table<K, V>,
);

}
```

This function takes the collection by value. If it contains no entries, it will be deleted, othewise the call will abort. `sui::table::Table` also has a convenience function,

```
module sui::table {

public fun drop<K: copy + drop + store, V: drop + store>(
    table: Table<K, V>,
);

}
```

> that can only be called for tables where the both the key *and* value type have `drop` which allows it to delete tables whether they are empty or not.

Reviewer: I find the wording here just a bit confusing, as the only new bit is the value part. The `K` type always needs `drop`

> that can only be called for tables where the both the key *and* value type have `drop` which allows it to delete tables whether they are empty or not.

Reviewer: ```suggestion
that can only be called for tables where the value type also has `drop`, which allows it to delete tables whether they are empty or not.
```
